File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
while(True):
    st=[]
    answer=True
    str=sys.stdin.readline()
    str=str.rstrip('\n')
    if(str=='.'):
        break
    for i in str:
        if(i=='('):
            st.append(1)
        elif(i=='['):
            st.append(2)
        elif(i==')'):
            if(not st or st[-1]!=1):
                answer=False
                break
            st.pop()
        elif(i==']'):
            if(not st or st[-1]!=2):
                answer=False
                break
            st.pop()
    if(not answer or st):
        print("no")
    else:
        print("yes")

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <stdio.h>
#include <iostream>
#include <string>
#include <stack>
#include <utility>
#include <algorithm>
int main(){
    std::cin.tie(0);
    std::ios::sync_with_stdio(0);
    std::stack <int> st{};
    std::string str{};
    bool answer=true;
    while(true){
        st={};
        answer=true;
        std::getline (std::cin, str);
        if(str=="."){
            break;
        }
        for(char i: str){
            if(i=='('){
                st.push(1);
            }
            else if(i=='['){
                st.push(2);
            }
            else if(i==')'){
                if(st.empty() || st.top()!=1){answer=false; break;}
                    st.pop();
            }
            else if(i==']'){
                if(st.empty() || st.top()!=2){answer=false; break;}
                    st.pop();
                }
            }
        if(st.empty() && answer){
            std::cout << "yes\n";
        }
        else{
            std::cout << "no\n";
        }
    }
}